src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -3941,3 +4086,363 @@ CompareInsertValuesByShardId(const void *leftElement, const void *rightElement)
 		}
 	}
 }
+
+
+/*
+ * IsMergeAllowedOnRelation takes a relation entry and checks if MERGE command is
+ * permitted on special relations, such as materialized view, returns true only if
+ * it's a "source" relation.
+ */
+bool
+IsMergeAllowedOnRelation(Query *parse, RangeTblEntry *rte)
+{
+	if (!IsMergeQuery(parse))
+	{
+		return false;
+	}
+
+	RangeTblEntry *targetRte = rt_fetch(parse->resultRelation, parse->rtable);
+
+	/* Is it a target relation? */
+	if (targetRte->relid == rte->relid)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * ErrorIfDistTablesNotColocated Checks to see if
+ *
+ *   - There are a minimum of two distributed tables (source and a target).
+ *   - All the distributed tables are indeed colocated.
+ *   - MERGE relations are joined on the distribution column
+ *          MERGE .. USING .. ON target.dist_key = source.dist_key
+ *
+ * If any of the conditions are not met, it raises an exception.
+ */
+static DeferredErrorMessage *
+ErrorIfDistTablesNotColocated(Query *parse, List *distTablesList,
+							  PlannerRestrictionContext *plannerRestrictionContext)
+{
+	/* All MERGE tables must be distributed */
+	if (list_length(distTablesList) < 2)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "For MERGE command, both the source and target "
+							 "must be distributed", NULL, NULL);
+	}
+
+	/* All distributed tables must be colocated */
+	if (!AllRelationsInListColocated(distTablesList, RANGETABLE_ENTRY))
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "For MERGE command, all the distributed tables "
+							 "must be colocated", NULL, NULL);
+	}
+
+	/* Are source and target tables joined on distribution column? */
+	if (!RestrictionEquivalenceForPartitionKeys(plannerRestrictionContext))
```

Comment:
The name of the function sounds too strict definition of "colocation". 

Conceptually, we mostly refer "colocation" as colocationId's in the code. Whereas RestrictionEquivalenceForPartitionKeys() check distribution key equality. So, I'd probably adjust the function name to reflect the difference.

Reply:
Yes, it checks for two conditions, changed it to `ErrorIfTablesNotColocatedAndJoinedOnDistColumn()`